pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,628 @@
+	case ipPoolUpdate:
+		// if IPPool update, alive nodes may leave or join, enqueue egresses
+		ipPool := obj.(*v1alpha2.ExternalIPPool)
+		c.syncConHashByIPPool(ipPool)
+		c.notify(ipPool.Name, NotifyEventTypeIPPoolChanged)
+	case ipPoolDelete:
+		// if IPPool delete, all the egresses has deleted, refs https://github.com/antrea-io/antrea/pull/2237
+		// just delete the conHash
+		ipPool := obj.(*v1alpha2.ExternalIPPool)
+		delete(c.conHashMap, ipPool.Name)
+	case nodeLabelUpdate:
+		// node alive in cluster, and may leave from a IPPool or join a IPPool
+		// all egresses with a IPPool would enqueue
+		poolNames := obj.(sets.String)
+		for _, poolName := range poolNames.List() {
+			ipPool, _ := c.ipPoolLister.Get(poolName)
+			c.syncConHashByIPPool(ipPool)
+			c.notify(ipPool.Name, NotifyEventTypeIPPoolChanged)
+		}
+	}
+}
+
+func (c *Cluster) syncConHashByIPPool(ipPool *v1alpha2.ExternalIPPool) {
+	nodeSel, _ := metav1.LabelSelectorAsSelector(&ipPool.Spec.NodeSelector)
+	nodes, _ := c.nodeLister.List(nodeSel)
+	conHash, ok := c.conHashMap[ipPool.Name]
+	if !ok {
+		conHash = newNodeConsistentHashMap()
+	}
+	aliveNodes := c.aliveNodeList()
+	for _, node := range nodes {
+		if aliveNodes.Has(node.Name) {
+			conHash.Add(node.Name)
+		}
+	}
+	c.conHashMap[ipPool.Name] = conHash
+}
+
+func (c *Cluster) syncConsistentHashMap(nodeEvent *memberlist.NodeEvent) {
+	c.conHashMapRWLock.Lock()
+	defer c.conHashMapRWLock.Unlock()
+	ipPools, _ := c.ipPoolLister.List(labels.Everything())
+	switch node, event := nodeEvent.Node, nodeEvent.Event; event {
+	case memberlist.NodeJoin:
+		node, err := c.nodeLister.Get(node.Name)
+		if err != nil {
+			return
+		}
+		for _, ipPool := range ipPools {
+			conHash, ok := c.conHashMap[ipPool.Name]
+			if !ok {
+				conHash = newNodeConsistentHashMap()
+			}
+			if c.matchNodeSelector(node, ipPool) {
+				klog.InfoS("Node event: join node", "node name", node.String())
+				conHash.Add(node.Name)
+			}
+		}
+		c.notify(node.Name, NotifyEventTypeNodeJoined)
+	case memberlist.NodeLeave:
+		for _, ipPool := range ipPools {
+			klog.InfoS("Node event: leave node", "node name", node.String())
+			c.syncConHashByIPPool(ipPool)
+		}
+		c.notify(node.Name, NotifyEventTypeNodeLeave)
+	default:
+		klog.InfoS("Node event: update node", "node name", node.String())
+	}
+}
+
+func (c *Cluster) updateLocalNodeStatus() {
+	c.localNodeEventCh <- true
+}
+
+// aliveNodeList return alive nodeName list in cluster
+func (c *Cluster) aliveNodeList() sets.String {
+	aliveMembers := c.mList.Members()
+	nodes := make([]string, len(aliveMembers))
+	for i, node := range aliveMembers {
+		nodes[i] = node.Name
+	}
+	return sets.NewString(nodes...)
+}
+
+// LocalNodeJoined if merbers num in cluster is 1 means local node not joined in other clusters
+func (c *Cluster) LocalNodeJoined() bool {
+	// while node joined cluster, num of member will >= 1; if leave cluster, num of member will be 0
+	return c.mList.NumMembers() >= 1
```

Comment:
what if this is single node cluster? Doesn't memberlist have a method to check whether this node joins others successfully?

Reply:
there is a method I have tried but it sames does not work well, memberlist.LocalNode().State is always alive whether node leave or join in cluster.
when a node left cluster, memberlist.NumMembers() will return 0, but the local node state memberlist.LocalNode().State is still alive, I have a little doubts here

Reply:
I think there are two cases: only this node is selected; this node and at least one other node is selected. In the former case, LocalNodeJoined is always true. In the latter case, LocalNodeJoined should be determined by whether it ever joins other node successfully.
I believe the return value of `Join` and the nodeEvent can tell the latter.

Anyway, can you consider https://github.com/antrea-io/antrea/pull/2186#discussion_r653567560 first? I can see dynamic joining makes the code complex and error prone. Can we simplify it in the first version?

Reply:
done